src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,50 +1,162 @@
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = findByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = findByIndex(index);
+        remove(node);
+        return node.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        if (isEmpty()) {
+            throw new IndexOutOfBoundsException("Cannot delete from empty list");
+        }
+        try {
+            Node<T> node = findByValue(object);
+            remove(node);
+        } catch (NoSuchElementException e) {
+            return false;
+        }
+        return true;
+    }
+
+    private void remove(Node<T> node) {
+        if (node.previous == null) {
+            head = node.next;
+        } else {
+            node.previous.next = node.next;
+        }
+        if (node.next == null) {
+            tail = node.previous;
+        } else {
+            node.next.previous = node.previous;
+        }
+        size--;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private class Node<T> {
+        private Node<T> previous;
+        private T value;
+        private Node<T> next;
+
+        public Node(Node<T> previous, T value, Node<T> next) {
+            this.previous = previous;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private Node<T> findByIndex(int index) {
+        Node<T> node;
+        if (index <= size / 2) {
+            node = head;
+            int listIndex = 0;
+            while (listIndex != index) {
+                node = node.next;
+                listIndex++;
+            }
+        } else {
+            node = tail;
+            int listIndex = size - 1;
+            while (listIndex != index) {
+                node = node.previous;
+                listIndex--;
+            }
+        }
+        return node;
+    }
+
+    private boolean checkIndex(int index) {
+        if (index >= 0 && index < size) {
+            return true;
+        }
+        throw new IndexOutOfBoundsException("Please check index");
+    }
+
+    private Node<T> findByValue(T value) {
+        Node<T> node = head;
+        while (node != null) {
+            if (node.value == value || (node.value != null && node.value.equals(value))) {
+                return node;
+            }
+            node = node.next;
+        }
+        throw new NoSuchElementException("No such element in list");
```

Comment:
we don't need exception here, if you use it for remove just return boolean and you won't need to use try catch